### ISSUE TITLE
ScenarioList and ExecuteList management are not robust to several parallel sessions.

### ISSUE THREAD

**Issue by danielolsen** [Breakthrough-Energy/PowerSimData#426]
I've seen a few different errors that all seem to originate from some error to read/parse the ScenarioList, which happens sometimes if I'm running commands in several separate python processes at the same time.

Traceback example:
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
c:\python38\lib\site-packages\powersimdata\input\grid.py in __init__(self, interconnect, source, engine)
     32         try:
---> 33             self.model_immutables = ModelImmutables(source)
     34         except ValueError:

c:\python38\lib\site-packages\powersimdata\network\model.py in __init__(self, model)
     11         """Constructor."""
---> 12         self._check_model(model)
     13         self.model = model

c:\python38\lib\site-packages\powersimdata\network\model.py in _check_model(model)
     33         if model not in possible:
---> 34             raise ValueError("model must be one of %s" % " | ".join(possible))
     35 

ValueError: model must be one of usa_tamu

During handling of the above exception, another exception occurred:

EmptyDataError                            Traceback (most recent call last)
<ipython-input-3-388746558d76> in <module>
      1 # Fetch the scenario, pass to the ScenarioInfo object
----> 2 scenario = Scenario(scenario_num)

c:\python38\lib\site-packages\powersimdata\scenario\scenario.py in __init__(self, descriptor)
     38                     self.state = Execute(self)
     39                 elif state == "analyze":
---> 40                     self.state = Analyze(self)
     41             except AttributeError:
     42                 return

c:\python38\lib\site-packages\powersimdata\scenario\analyze.py in __init__(self, scenario)
     37 
     38         self._set_allowed_state()
---> 39         self._set_ct_and_grid()
     40 
     41     def _set_allowed_state(self):

c:\python38\lib\site-packages\powersimdata\scenario\analyze.py in _set_ct_and_grid(self)
     48         input_data = InputData(data_loc=self.data_loc)
     49         grid_mat_path = input_data.get_data(self._scenario_info, "grid")
---> 50         self.grid = Grid(
     51             interconnect=[None],
     52             source=grid_mat_path,

c:\python38\lib\site-packages\powersimdata\input\grid.py in __init__(self, interconnect, source, engine)
     34         except ValueError:
     35             self.model_immutables = ModelImmutables(
---> 36                 _get_grid_model_from_scenario_list(source)
     37             )
     38 

c:\python38\lib\site-packages\powersimdata\input\grid.py in _get_grid_model_from_scenario_list(source)
    163     scenario_number = int(os.path.basename(source).split("_")[0])
    164     slm = ScenarioListManager(Context.get_data_access())
--> 165     return slm.get_scenario(scenario_number)["grid_model"]

c:\python38\lib\site-packages\powersimdata\data_access\scenario_list.py in get_scenario(self, descriptor)
    118             print("------------------")
    119 
--> 120         table = self.get_scenario_table()
    121         try:
    122             matches = table.index.isin([int(descriptor)])

c:\python38\lib\site-packages\powersimdata\data_access\scenario_list.py in get_scenario_table(self)
     93         :return: (*pandas.DataFrame*) -- scenario list as a data frame.
     94         """
---> 95         return self.get_table()
     96 
     97     def generate_scenario_id(self):

c:\python38\lib\site-packages\powersimdata\data_access\csv_store.py in get_table(self)
     51 
     52         if local_path.is_file():
---> 53             return self._parse_csv(local_path)
     54         else:
     55             raise FileNotFoundError(f"{filename} does not exist locally.")

c:\python38\lib\site-packages\powersimdata\data_access\csv_store.py in _parse_csv(self, file_object)
     62         :return: (*pandas.DataFrame*) -- the specified file as a data frame.
     63         """
---> 64         table = pd.read_csv(file_object)
     65         table.set_index("id", inplace=True)
     66         table.fillna("", inplace=True)

c:\python38\lib\site-packages\pandas\io\parsers.py in read_csv(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)
    608     kwds.update(kwds_defaults)
    609 
--> 610     return _read(filepath_or_buffer, kwds)
    611 
    612 

c:\python38\lib\site-packages\pandas\io\parsers.py in _read(filepath_or_buffer, kwds)
    460 
    461     # Create the parser.
--> 462     parser = TextFileReader(filepath_or_buffer, **kwds)
    463 
    464     if chunksize or iterator:

c:\python38\lib\site-packages\pandas\io\parsers.py in __init__(self, f, engine, **kwds)
    817             self.options["has_index_names"] = kwds["has_index_names"]
    818 
--> 819         self._engine = self._make_engine(self.engine)
    820 
    821     def close(self):

c:\python38\lib\site-packages\pandas\io\parsers.py in _make_engine(self, engine)
   1048             )
   1049         # error: Too many arguments for "ParserBase"
-> 1050         return mapping[engine](self.f, **self.options)  # type: ignore[call-arg]
   1051 
   1052     def _failover_to_python(self):

c:\python38\lib\site-packages\pandas\io\parsers.py in __init__(self, src, **kwds)
   1896 
   1897         try:
-> 1898             self._reader = parsers.TextReader(self.handles.handle, **kwds)
   1899         except Exception:
   1900             self.handles.close()

pandas\_libs\parsers.pyx in pandas._libs.parsers.TextReader.__cinit__()

EmptyDataError: No columns to parse from file
```

**Comment by danielolsen** [comment]
I managed to corrupt the ScenarioList.csv file on the server when I encountered this error while creating a scenario (luckily recovered via backups). This may be more urgent that we originally thought.

**Comment by rouille** [comment]
> I managed to corrupt the ScenarioList.csv file on the server when I encountered this error while creating a scenario (luckily recovered via backups). This may be more urgent that we originally thought.

Where you creating multiple scenarios at the same time?

**Comment by danielolsen** [comment]
> > I managed to corrupt the ScenarioList.csv file on the server when I encountered this error while creating a scenario (luckily recovered via backups). This may be more urgent that we originally thought.
> 
> Where you creating multiple scenarios at the same time?

No, I was creating one scenario while analyzing another.

**Comment by danielolsen** [comment]
@jon-hagg @rouille and I brainstormed a bit, we are pretty sure that the root cause is that the local copy of the ScenarioList can get corrupted by a partial download from one process while another process is trying to read/parse it (first issue), or it can get corrupted by a partial download from one process while another process is trying to push it to the server (second issue). When the connection to the server is slow, a script may be downloading the ScenarioList several times and the time spent in `sftp.get` can be the majority of the script execution time, which poses a high risk when we have several processes working concurrently.

To address this, we've got the following ideas:
- When downloading, download to a tempfile location and then move to the local ScenarioList location (should help avoid local corruption for the first issue).
- When adding to the scenario list and uploading, save the updated file at a tempfile location and then upload to the server (should help avoid local corruption in the second issue which becomes server file corruption).
- Optionally, add an optional parameter to the `get_scenario_table` method of the ScenarioListManager which prioritizes looking in the cached local copy before trying to get the latest from the server (should help reduce the number of download calls to the server).
